File: app/src/main/java/com/codepath/shapes/MainActivity.java
package com.codepath.shapes;

import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.annotation.ColorInt;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import com.codepath.shapes.databinding.ActivityMainBinding;
import com.codepath.shapes.io.ShapeSerializer;
import com.codepath.shapes.shape.Circle;
import com.codepath.shapes.shape.Rectangle;
import com.codepath.shapes.shape.Shape;
import com.codepath.shapes.shape.Text;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private ActivityMainBinding mBinding;
    private SelectShape mSelectedShape = SelectShape.RECTANGLE;

    @ColorInt
    private int mRectColor;

    @ColorInt
    private int mCircleColor;

    @ColorInt
    private int mTextColor;


    private float mRectangleSide;
    private float mCircleRadius;
    private float mTextSize;

    private ShapeSerializer mShapeSerializer;

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        mRectColor = getResources().getColor(R.color.colorRectangle);
        mCircleColor = getResources().getColor(R.color.colorCircle);
        mTextColor = getResources().getColor(R.color.colorText);
        mRectangleSide = getResources().getDimensionPixelOffset(R.dimen.rectangleSide);
        mCircleRadius = getResources().getDimensionPixelOffset(R.dimen.circleRadius);
        mTextSize = getResources().getDimension(R.dimen.textFontSize);

        mShapeSerializer = new ShapeSerializer(this);

        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        setSupportActionBar(mBinding.toolbar);

        mBinding.screen.init();
        mBinding.screen.setOnTapListener(new Screen.OnTapListener() {
            @Override
            public void onTap(float x, float y) {
                final Shape shape;
                switch (mSelectedShape) {
                    case RECTANGLE:
                        shape = new Rectangle(x, y, mRectColor, mRectangleSide, mRectangleSide);
                        break;

                    case CIRCLE:
                        shape = new Circle(x, y, mCircleColor, mCircleRadius);
                        break;

                    case TEXT:
                        shape = new Text(x, y, mTextColor, mTextSize, "Kotlin");
                        break;

                    default:
                        //Hey Java, this shouldn't happen
                        shape = new Rectangle(0, 0, 0, 0, 0);
                }

                mBinding.screen.addShape(shape);
            }
        });

        mBinding.bottomBar.setOnNavigationItemSelectedListener(new BottomNavigationView.OnNavigationItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.action_circle:
                        mSelectedShape = SelectShape.CIRCLE;
                        break;
                    case R.id.action_rectangle:
                        mSelectedShape = SelectShape.RECTANGLE;
                        break;
                    case R.id.action_text:
                        mSelectedShape = SelectShape.TEXT;
                        break;
                }
                return true;
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        final MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.toolbar_options, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_erase:
                mBinding.screen.clear();
                return true;

            case R.id.action_save:
                mShapeSerializer.saveShapes(mBinding.screen.getShapeList(), new ShapeSerializer.OnSaveFinishedCallback() {
                    @Override
                    public void onSaveFinished() {
                        Snackbar.make(mBinding.getRoot(), "Saved", Snackbar.LENGTH_SHORT).show();
                    }
                });
                return true;

            case R.id.action_restore:
                mShapeSerializer.restoreShapes(new ShapeSerializer.OnRestoreFinishedCallback() {
                    @Override
                    public void onRestoreFinished(@NonNull List<Shape> shapeList) {
                        Snackbar.make(mBinding.getRoot(), "Restored", Snackbar.LENGTH_SHORT).show();
                        mBinding.screen.setShapeList(shapeList);
                    }
                });
                return true;

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private enum SelectShape {
        RECTANGLE, CIRCLE, TEXT
    }
}

File: app/src/main/java/com/codepath/shapes/MainActivity.kt
package com.codepath.shapes

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.annotation.ColorRes
import android.support.annotation.DimenRes
import android.support.design.widget.Snackbar
import android.support.v7.app.AppCompatActivity
import android.view.Menu
import android.view.MenuItem
import com.codepath.shapes.databinding.ActivityMainBinding
import com.codepath.shapes.io.ShapeSerializer
import com.codepath.shapes.shape.Circle
import com.codepath.shapes.shape.Rectangle
import com.codepath.shapes.shape.Text

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private var selectedShape = SelectShape.RECTANGLE
    private val shapeSerializer = ShapeSerializer(this)

    private val rectColor by lazyColor(R.color.colorRectangle)
    private val circleColor by lazyColor(R.color.colorCircle)
    private val textColor by lazyColor(R.color.colorText)

    private val rectangleSide by lazyDimension(R.dimen.rectangleSide)
    private val circleRadius by lazyDimension(R.dimen.circleRadius)
    private val textSize by lazyFontSize(R.dimen.textFontSize)

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main).apply {
            setSupportActionBar(toolbar)

            screen.tapListener = { x, y ->
                val shape = when (selectedShape) {
                    SelectShape.RECTANGLE -> Rectangle(x, y, rectColor, rectangleSide, rectangleSide)
                    SelectShape.CIRCLE -> Circle(x, y, circleColor, circleRadius)
                    SelectShape.TEXT -> Text(x, y, textColor, textSize, "Kotlin")
                }
                screen.addShape(shape)
            }

            bottomBar.setOnNavigationItemSelectedListener { item ->
                selectedShape = when (item.itemId) {
                    R.id.action_circle -> SelectShape.CIRCLE
                    R.id.action_rectangle -> SelectShape.RECTANGLE
                    R.id.action_text -> SelectShape.TEXT
                    else -> throw IllegalArgumentException("Unexpected selected value ${item.itemId}")
                }
                true
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.toolbar_options, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_erase -> {
            binding.screen.clear()
            true
        }
        R.id.action_save -> {
            shapeSerializer.saveShapes(binding.screen.shapeList) {
                Snackbar.make(binding.root, "Saved", Snackbar.LENGTH_SHORT).show()
            }
            true
        }
        R.id.action_restore -> {
            shapeSerializer.restoreShapes { shapeList ->
                binding.screen.shapeList = shapeList
                Snackbar.make(binding.root, "Restored", Snackbar.LENGTH_SHORT).show()
            }
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    private enum class SelectShape {
        RECTANGLE, CIRCLE, TEXT
    }

    private fun lazyColor(@ColorRes color: Int) = lazy({ resources.getColor(color) })
    private fun lazyDimension(@DimenRes dimenRes: Int) = lazy({ resources.getDimensionPixelOffset(dimenRes).toFloat() })
    private fun lazyFontSize(@DimenRes dimenRes: Int) = lazy({ resources.getDimension(dimenRes) })
}
